File: domain/QuickCapture.kt
package it.notes.ecosystem.domain

/** Plain text only: URLs remain text and are never fetched. */
data class CaptureSeed(val title: String = "", val body: String = "", val checklist: Boolean = false) {
    val hasContent: Boolean get() = title.isNotBlank() || body.isNotBlank()
}
object QuickCapture {
    const val MAX_BYTES = 200 * 1024
    fun shared(text: String?, subject: String?): CaptureSeed {
        val body = text.orEmpty()
        val title = subject.orEmpty().trim()
        require(title.length <= 8000) { "Il titolo condiviso è troppo lungo." }
        require(body.length <= MAX_BYTES && title.length + body.length <= MAX_BYTES) { "Testo condiviso oltre 200 KB." }
        require(title.toByteArray(Charsets.UTF_8).size + body.toByteArray(Charsets.UTF_8).size <= MAX_BYTES) { "Testo condiviso oltre 200 KB." }
        require('\u0000' !in body && '\u0000' !in title) { "Il contenuto non è testo valido." }
        return CaptureSeed(title, body).also { require(it.hasContent) { "Nessun testo da aggiungere. Condividi un testo o un link." } }
    }
}

File: flutter_app/lib/src/domain/quick_capture.dart
import 'dart:convert';

class CaptureSeed {
  const CaptureSeed({
    this.title = '',
    this.body = '',
    this.checklist = false,
  });

  final String title;
  final String body;
  final bool checklist;

  bool get hasContent => title.trim().isNotEmpty || body.trim().isNotEmpty;
}

abstract final class QuickCapture {
  static const maxBytes = 200 * 1024;

  static CaptureSeed shared(String? text, String? subject) {
    final body = text ?? '';
    final title = (subject ?? '').trim();
    if (title.length > 8000) {
      throw const FormatException('Il titolo condiviso è troppo lungo.');
    }
    if (utf8.encode(title).length + utf8.encode(body).length > maxBytes) {
      throw const FormatException('Testo condiviso oltre 200 KB.');
    }
    if (title.contains('\u0000') || body.contains('\u0000')) {
      throw const FormatException('Il contenuto non è testo valido.');
    }
    final seed = CaptureSeed(title: title, body: body);
    if (!seed.hasContent) {
      throw const FormatException(
        'Nessun testo da aggiungere. Condividi un testo o un link.',
      );
    }
    return seed;
  }
}

class SharedCaptureFile {
  const SharedCaptureFile({
    required this.path,
    required this.name,
    required this.mime,
  });

  final String path;
  final String name;
  final String mime;

  factory SharedCaptureFile.fromMap(Map<Object?, Object?> map) =>
      SharedCaptureFile(
        path: map['path']?.toString() ?? '',
        name: map['name']?.toString() ?? 'Allegato',
        mime: map['mime']?.toString() ?? '',
      );
}

class IncomingCapture {
  const IncomingCapture({
    required this.seed,
    this.files = const [],
  });

  final CaptureSeed seed;
  final List<SharedCaptureFile> files;

  bool get empty => !seed.hasContent && files.isEmpty;

  factory IncomingCapture.fromMap(Map<Object?, Object?> map) {
    final rawFiles = map['files'];
    return IncomingCapture(
      seed: CaptureSeed(
        title: map['title']?.toString() ?? '',
        body: map['body']?.toString() ?? '',
        checklist: map['checklist'] == true,
      ),
      files: rawFiles is List
          ? rawFiles
              .whereType<Map>()
              .map(
                (item) => SharedCaptureFile.fromMap(
                  item.cast<Object?, Object?>(),
                ),
              )
              .toList(growable: false)
          : const [],
    );
  }
}
